README.md
<div align="center">

![Banner](/images/editor_banner.jpg)
----
[![CI](https://github.com/Rosemoe/CodeEditor/actions/workflows/gradle.yml/badge.svg?event=push)](https://github.com/Rosemoe/CodeEditor/actions/workflows/gradle.yml)
[![GitHub license](https://img.shields.io/github/license/Rosemoe/CodeEditor)](https://github.com/Rosemoe/CodeEditor/blob/main/LICENSE)
[![Maven Central](https://img.shields.io/maven-central/v/io.github.Rosemoe.sora-editor/editor.svg?label=Maven%20Central)]((https://search.maven.org/search?q=io.github.Rosemoe.sora-editor%20editor))   
[![Telegram](https://img.shields.io/badge/Join-Telegram-blue)](https://t.me/rosemoe_code_editor)
[![QQ](https://img.shields.io/badge/Join-QQ_Group-ff69b4)](https://jq.qq.com/?_wv=1027&k=n68uxQws)

sora-editor is a cool and optimized code editor on Android platform

</div>

Read this in other languages: [English](README.md), [简体中文](README.zh-cn.md), [Español](README.es.md).

***Work In Progress*** This project is still developing slowly.   
Download the newest sources from [Releases](https://github.com/Rosemoe/CodeEditor/releases)
instead of cloning this repository directly.
**Issues and pull requests are welcome.**

## Features

- [x] Syntax highlighting
- [x] Auto-completion (with [code snippets](https://macromates.com/manual/en/snippets))
- [x] Auto indent
- [x] Code block lines
- [x] Scale text
- [x] Undo/redo
- [x] Search and replace
- [x] Auto wordwrap
- [x] Show non-printable characters
- [x] Error/Warning/Typo/Deprecated indicators
- [x] Text magnifier
- [x] Incremental highlight analysis
- [x] Highlight bracket pairs
- [x] Event System

## Key bindings

When working with a physical keyboard, you can use key bindings for performing various text
actions.
The editor provides support for some key bindings by default.
However, you can subscribe
to [`KeyBindingEvent`](https://github.com/Rosemoe/sora-editor/blob/main/editor/src/main/java/io/github/rosemoe/sora/event/KeyBindingEvent.java)
and add your own key bindings. You can even override the default key bindings and perform actions of
your own.

The currently supported key bindings are mostly similar to Android Studio/Intellij IDEA.
See the [supported key bindings](./keybindings.md).

## Screenshots

<div style="overflow: hidden">
<img src="/images/general.jpg" alt="GeneralAppearance" width="40%" align="bottom" />
<img src="/images/problem_indicators.jpg" alt="ProblemIndicator" width="40%" align="bottom" />
</div>

## Get started

Add to your app's dependencies:

```Gradle
dependencies {
    implementation(platform("io.github.Rosemoe.sora-editor:bom:<versionName>"))
    implementation("io.github.Rosemoe.sora-editor:<moduleName>")
}
```

Available modules:

- editor   
  Widget library containing all basic things of the framework
- editor-lsp   
  A convenient library for creating languages by using Language Server Protocol (aka LSP)
- language-java   
  A simple implementation for Java highlighting and identifier auto-completion
- language-textmate   
  An advanced highlighter for the editor. You can find textmate language bundles and themes and load
  them by using this
  module. The internal implementation of textmate is from [tm4e](https://github.com/eclipse/tm4e).
- language-treesitter   
  Offer [tree-sitter](https://tree-sitter.github.io/tree-sitter/) support for editor. This can be used to 
  parse the code to an AST fast and incrementally, which is helpful for accurate highlighting and providing completions.
  Note that this module only provides incremental paring and highlighting. Thanks to Java bindings [android-tree-sitter](https://github.com/AndroidIDEOfficial/android-tree-sitter/)

Check the newest version from the badge above
or [Releases](https://github.com/Rosemoe/CodeEditor/releases).

## Discuss

* Official QQ Group:[216632648](https://jq.qq.com/?_wv=1027&k=n68uxQws)
* Official [Telegram Group](https://t.me/rosemoe_code_editor)

## Contributors

<a href="https://github.com/Rosemoe/sora-editor/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=Rosemoe/sora-editor" />
</a>

## License

```
sora-editor - the awesome code editor for Android
https://github.com/Rosemoe/sora-editor
Copyright (C) 2020-2023  Rosemoe

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
USA

Please contact Rosemoe by email 2073412493@qq.com if you need
additional information or have any questions
```

## Acknowledgements

Thanks to [JetBrains](https://www.jetbrains.com/?from=CodeEditor) for allocating free open-source
licences for IDEs such as [IntelliJ IDEA](https://www.jetbrains.com/idea/?from=CodeEditor).   
[<img src=".github/jetbrains-variant-3.png" width="200"/>](https://www.jetbrains.com/?from=CodeEditor)


docs/changelog.md
# Changelog

## **[0.21.0](https://github.com/Rosemoe/sora-editor/releases/tag/0.21.0) (2023-01-08)**

> This release includes bug fixes and little improvement as well as minor API changes.

### Bugs fixed

- incorrect horizontal scroll range when `renderFunctionCharacter` is on
- `TSQuery` is not verified before making access to it
- IME that relies on `InputConnection#getSurroundingText` gets invalid position description
  when `maxIPCTextLength` is
  exceeded
- invalid composing text range can be set through `EditorInputConnection#setComposingRegion`
- text becomes dirty when using old Gboard to delete characters fast
- animated row background is on wrong layer
- empty text is matched repeatly in regex, leading to OOM
- `StringIndex OutOfBoundsException` in `TextMateNewlineHandler` (by *
  *[@dingyi222666](https://github.com/dingyi222666)**)
- `CodeEditor#release` does not detach `EditorColorScheme`
- NPE when sending message in `AsyncIncrementalAnalyzeManager`
- leaking thread by `TextMateLanguage#updateLanguage`
- `ReplaceAction` can be added without its `DeleteAction`
- nested undo/redo is possible
- `JavaTextTokenizer` does not take down current
  token [#349](https://github.com/Rosemoe/sora-editor/issues/349)
- `AsyncIncrementalAnalyzeManager` does not update the spans of last line being affected by
  insertion [#350](https://github.com/Rosemoe/sora-editor/issues/350)
- `Styles#eraseAllLineStyles` raises npe when no line style is there

### Improvments

- better search experience, including support for whole word search, speed improvement of regex
  search, cyclic jumping
  and better scrolling strategy [#321](https://github.com/Rosemoe/sora-editor/issues/321)
- performance improvement of tree-sitter predicates
- avoid edge effect in `CodeEditor#ensurePositionVisible`
- boost the speed of querying search results by binary search
- show built-in text actions window on insert handle release
- add touch slop before magnifier is triggered
- add `I18nConfig` for application provided replacement string resources
- add `DirectAccessProps#clipboardTextLengthLimit` and add tip when text is unable to be copied due
  to limit
  or `TransactionTooLargeException`
- more exact position of built-in text actions window
- editor uses global default color scheme on startup
- enhanced function of `DirectAccessProps#disallowSuggestions` (tested on newest Gboard and Sogou
  Input for MIUI)
- adapt view parameters for long screenshot in MIUI system
- add two new accessibility actions since Android M

### Updated API

- new function `Content#substring`
- new event `PublishSearchResultEvent`, which is called when search result is available in main
  thread or searching is
  stopped
- new `I18nConfig` class for replacing string resources
- new function `add CodeEditor#isAntiWordBreaking`
- **[BREAKING]** new API in `EditorColorScheme` for global default theme. All newly-created editors
  use the global
  default theme. Your modifications to global default theme will reflect in those editors.
- **[BREAKING]** `CodeEditor#getScroller` now has return type `EditorScroller`. Get the
  original `OverScroller`
  by `EditorScroller#getImplScroller`
- **[BREAKING]** `AsyncIncrementalAnalyzeManager` and `TsAnalyzeManager` will receive thread
  interruption on exit

## **[0.20.4](https://github.com/Rosemoe/sora-editor/releases/tag/0.20.4) (2022-12-30)**

### Bugs Fixed

- member variable highlighting in tree-sitter
- editor crashes in rendering highlighted delimeters when spans are corrupted
- `EditorDiagnosticTootipWindow` can not be got from editor components
- `EventManager` can not be disabled

### New Features

- `CodeEditor#createSubEventManager()`

## **[0.20.2](https://github.com/Rosemoe/sora-editor/releases/tag/0.20.2) (2022-12-27)**

### Fixed Bugs

- indent is unexpectedly doubled in keybinding `Ctrl+D`
- default completion layout can not be clicked after animation unless you scroll the list
- invalid indices are used in editing `TSTree` in main thread
- `IndexOutOfBoundsException` when appending text if tree-sitter is used
- unable to run predicates after text editions in tree-sitter

### New features

- line info panel style is now a bubble [#283](https://github.com/Rosemoe/sora-editor/issues/283)
- `Content.getDocumentVersion()` is added for checking document modifications
- `DiagnosticDetail` for describing diagnostics and
  builtin `EditorDiagnosticTooltipWindow` [#314](https://github.com/Rosemoe/sora-editor/issues/314)
- experimental `Quickfix` item
- support members scope in tree-sitter (see `LocalsCaptureSpec` for more information) and improve
  speed in finding
  definitions
- custom text provider for line info panel `LineNumberTipTextProvider`

## **[0.20.1](https://github.com/Rosemoe/sora-editor/releases/tag/0.20.1) (2022-12-21)**

- Fixed a critical bug of tree-sitter code blocks in previous release of 0.20.0

## **[0.20.0](https://github.com/Rosemoe/sora-editor/releases/tag/0.20.0) (2022-12-21)**

!> Use `0.20.1` instead

### Bugs fixed

- memory leaks in `LanguageServerWrapper`
- `EditorSearcher.gotoPrevious()` always jumps to the first item when regex is enabled
- transparent line in selection handle
  drawable ([#270](https://github.com/Rosemoe/sora-editor/issues/270), fixed
  by [@tegajoel](https://github.com/tegajoel) [#312](https://github.com/Rosemoe/sora-editor/pull/312))
- whole document becomes black-colored when user is scaling in some languages
- `SimpleAnalyzeManager` ignores line separators
- `OnigRegExp` returns corrupted cache in multi-threaded
  access ([#315](https://github.com/Rosemoe/sora-editor/issues/315)
  by [@xyzxqs](https://github.com/xyzxqs))
- `TextReference.toString()` does not return string of backed sequence
- text context region can be unexpectedly bigger the bounds of
  text [#318](https://github.com/Rosemoe/sora-editor/issues/318)
- textmate newline handler fixs ([#319](https://github.com/Rosemoe/sora-editor/issues/319)

buildSrc/src/main/java/Versions.java
/*
 *    sora-editor - the awesome code editor for Android
 *    https://github.com/Rosemoe/sora-editor
 *    Copyright (C) 2020-2022  Rosemoe
 *
 *     This library is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU Lesser General Public
 *     License as published by the Free Software Foundation; either
 *     version 2.1 of the License, or (at your option) any later version.
 *
 *     This library is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public
 *     License along with this library; if not, write to the Free Software
 *     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *     USA
 *
 *     Please contact Rosemoe by email 2073412493@qq.com if you need
 *     additional information or have any questions
 */
@SuppressWarnings("unused")
public final class Versions {

    // Project versions
    public final static String versionName = "0.20.2";
    public final static int versionCode = 73;

    // Platform & Tool versions
    public final static String buildToolsVersion = "33.0.0";
    public final static int compileSdkVersion = 33;
    public final static int minSdkVersion = 21;
    public final static int minSdkVersionHighApi = 26;
    public final static int targetSdkVersion = 33;

    public final static String kotlinVersion = "1.7.20";

    // Dependency versions
    public final static String antlrVersion = "4.9.2";
    public final static String androidxAnnotationVersion = "1.5.0";
    public final static String lsp4jVersion = "0.17.0";
    public final static String tsBindingVersion = "1.0.10";


    // TextMate Dependency versions(from tm4e 0.4.2)
    public final static String gsonVersion = "2.9.1";
    public final static String jcodingsVersion = "1.0.57";
    public final static String joniVersion = "2.1.43";
    public final static String batikCssVersion = "1.14";
    public final static String batikUtilVersion = "1.14";
    public final static String xercesImplVersion = "2.12.2";
}